Proper deregistration of static types and objects

dorozhkin-stc opened the pull request:
It turns out on Windows static destructror should run on behalf the same library which run static constructor. For example, the objects registered by `torch_cuda_cpp.dll` should be properly deregistered by `torch_cuda_cpp.dll`, otherwise `torch_cpu.dll` will crash on unload.

Hopefully fixes #61111

Commit:
Proper deregistration of static types and objects

Hopefully this fixes possible crash when unload torch_cpu on Windows.
See https://github.com/pytorch/pytorch/issues/61111 for the details.

peterjc123 commented on torch/csrc/jit/runtime/profiling_record.cpp:
```diff
@@ -22,9 +22,16 @@ class ProfileRegistry {
+  size_t registerProfileNode(const std::function<bool(const Node*)>& func) {
     std::lock_guard<std::mutex> guard(mutex_);
-    registry_funcs_.push_back(func);
+    const auto key = registry_funcs_.size();
```
This is somehow dangerous to use the number of the dictionary as key. Suppose I call unregister and then register once, then the last function will be overridden.

dorozhkin-stc replied:
Thanks, I have come up with saner version which keeps increasing counter in a member field.

peterjc123 reviewed (commented):
Thanks, the changes look good from my perspective. But some CI tests on Linux didn't pass, maybe rebasing could help? BTW, maybe @ezyang @albanD can help to review this stuff.

ezyang commented:
@jamesr66a do you know who the right person to review this is?

dorozhkin-stc commented on torch/custom_class_detail.h:
```diff
@@ -187,6 +187,21 @@ class TORCH_API class_base {
+TORCH_API void deregisterCustomClassMethod(const c10::QualifiedName& name);
+
+template<typename Unused = void>
+void deregisterCustomClassMethodOnExit(c10::QualifiedName name) {
```
That was intended to take rvalue-only. In the newer version I have redesigned the deregistration facility and moved it to the `torch::class_` which is already a template.

dorozhkin-stc commented:
Rebased on recent master and cleaned up. Is there a windows build corresponding to my branch available online so I can double-check if the issue fixed?

ezyang commented:
@dorozhkin-stc I went ahead and triggered binary builds for this PR, although I don't know if the binaries are saved in a spot you can get at (@zhouzhuojie do you know?)

jamesr66a commented:
Does this have implications on custom class usage that should be mentioned in the documentation?

dorozhkin-stc commented:
> Does this have implications on custom class usage that should be mentioned in the documentation?

None known. The intended static deregistration on library unload (which is the reverse of the static registration on library load) is designed to run completely automatically and transparent to the user.

ezyang commented:
you can get the windows binaries on the "Artifacts" tab, see https://app.circleci.com/pipelines/github/pytorch/pytorch/354963/workflows/78d15278-8a89-47ad-96ee-2bbeeb3b8cb5/jobs/14962925/artifacts for example

jamesr66a reviewed (approved):
This seems fine to me. Adding @tugsbayasgalan @SplitInfinity @gmagogsfm as most proximate torchbind owners for additional review

ezyang commented:
@jamesr66a will you land?

jamesr66a commented:
@ezyang  sure

jamesr66a commented:
@dorozhkin-stc can you rebase? Some of the test jobs seem to be failing due to infrastructure errors

dorozhkin-stc commented:
After rebase there are still some job failures. Seems unrelated though (hard drive is out of free space on Windows machine and deadline timeout on the others).

gmagogsfm commented:
Looks good, the CI test failures seem unrelated to your change. Could you rebase to trigger a re-run so that real issues (if any) don't get masked by those unrelated failures? Thanks!

DBraun commented:
Is there an update on this? I would really like to use a newer version of libtorch, and I think this issue is preventing me.

jamesr66a commented:
@dorozhkin-stc it looks like there's still infra failures on CI here, can you rebase again?

ezyang commented:
bumping
